<a href="https://colab.research.google.com/github/Mohamed-Harish-M-123/Capstone-Project-IV/blob/main/Online_Retail_Customer_Segmentation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##**Project Name: Online Retail Customer Segmentation**
####**Project Type:** Unsupervised ML
####**Contribution:** Individual
####**Name:** MOHAMED HARISH

###**GitHub link:** https://github.com/Mohamed-Harish-M-123/Capstone-Project-IV

##**Business context:**
####In this project, our task is to identify major customer segments on a transactional data set which contains all the transactions occuring between 01/12/2010 and 09/12/2011 for a UK based and registered non-store online retail. The company mainly sells unique all-occasion gifts. Many customer of the company are wholesalers.

##**Problem statement:**
####The problem at hand is to develop an effective online retail customer segmentation model using unsupervised machine learning techniques. The goal is to identify distinct groups of customers based on their purchasing behaviors, preferences, and demographic information, in order to enable personalized marketing strategies and enhance the overall customer experience.

###**Objective:**
####The objective of this project is to develop a customer segmentation model that can effectively group customers based on their purchasing behavior and demographic attributes, enabling the online retailer to personalize marketing campaigns, optimize product recommendations, and improve customer satisfaction. By understanding the distinct needs and preferences of different customer segments, the retailer can tailor their strategies to enhance customer engagement, increase sales, and drive long-term customer loyalty.

###**Importing libraries**

In [1]:
# Import required libraries for this project

from google.colab import drive

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

import warnings
warnings.filterwarnings('ignore')


In [2]:
# Mount the drive and read the csv file which is our dataset
drive.mount("/content/drive")

path = "/content/drive/MyDrive/dataset/Online Retail.csv"

df = pd.read_csv(path)

Mounted at /content/drive


In [3]:
# View the dataset
df.head(3)

InvoiceNo StockCode                         Description  Quantity  \
0    536365    85123A  WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                 WHITE METAL LANTERN         6   
2    536365    84406B      CREAM CUPID HEARTS COAT HANGER         8   

    InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/10 8:26       2.55     17850.0  United Kingdom  
1  12/1/10 8:26       3.39     17850.0  United Kingdom  
2  12/1/10 8:26       2.75     17850.0  United Kingdom

In [4]:
# View the dataset info

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


In [5]:
# Check the shape
df.shape

(541909, 8)

In [6]:
# Check for duplicates

df.duplicated().sum()

5268

In [7]:
# Check for the null values

df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

###Dataset overview:
- Shape of our dataset is 8 columns 541909 rows.
- Datatypes of the columns are 2 floats, 1 int, 5 objects
- 5268 duplicate records
- Description and CustomerID has null values of 1454 and 135080.

In [8]:
# Create a copy of your original dataframe

customer_df = df.copy()

In [9]:
# Check unique values in your dataframe

for col in customer_df.columns:
  print("Unique values in ",col," is", customer_df[col].nunique())

Unique values in  InvoiceNo  is 25900
Unique values in  StockCode  is 4070
Unique values in  Description  is 4223
Unique values in  Quantity  is 722
Unique values in  InvoiceDate  is 23260
Unique values in  UnitPrice  is 1630
Unique values in  CustomerID  is 4372
Unique values in  Country  is 38


In [10]:
# Convert the datatype of Invoice Date to datetime.

customer_df['InvoiceDate'] = pd.to_datetime(customer_df['InvoiceDate'])

customer_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


In [12]:
# Extract month, day, and time from the Invoice Date column

customer_df['InvoiceMonth'] = customer_df['InvoiceDate'].dt.month

customer_df['InvoiceDay'] = customer_df['InvoiceDate'].dt.day_name()

customer_df['InvoiceTime'] = customer_df['InvoiceDate'].dt.time

customer_df.head(5)

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  InvoiceMonth  \
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom            12   
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom            12   
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom            12   
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom            12   
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom            12   

  InvoiceDay InvoiceTime  
0  Wednesday    08:26:00  
1  Wednesday    08:26:00  
2  Wednesday    08:26:00  
3  Wednesday    08:26:00  
4  Wednesday    08:26:00